src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/StorePal.macOS.LoaderPal.cs

```diff
@@ -117,6 +117,18 @@ public void MoveTo(X509Certificate2Collection collection)
                             newPal = pal.MoveToKeychain(_keychain, safeSecKeyRefHandle) ?? pal;
                         }
 
+                        if (_keychain is SafeTemporaryKeychainHandle)
+                        {
+                            // If we used temporary keychain we need to prevent deletion.
+                            // on 10.15+ if keychain is unlinked, certain certificate operations may fail.
+                            bool success = false;
+                            _keychain.DangerousAddRef(ref success);
```

Comment:
I believe this makes the "temporary" keychain more or less "permanent", because all calls to Dispose, collectively, only call DangerousRelease once.

You'd need to tell the newPal that it should DangerousRelease instead of Dispose in order to get it to ever tick down to 0 and actually release; and then have the collection import call Dispose() on the temp handle here (after the loop ends).

Reply:
> I believe this makes the "temporary" keychain more or less "permanent", because all calls to Dispose, collectively, only call DangerousRelease once.
> You'd need to tell the newPal that it should DangerousRelease instead of Dispose in order to get it to ever tick down to 0 and actually release; 

I was about to agree with you but then I went to compare it with my observations:
- The `DangerousAddRef `/`Dispose` pair is a pre-existing construct introduced in #41787.
- I definitely do see `AppleCryptoNative_SecKeychainDelete` getting called and that's called only from the final release of the temporary keychain. This can be verified by taking [the repro from the issue](https://github.com/dotnet/runtime/issues/80176#issuecomment-1432194589) and adding `certificateToImport.Dispose()` at the end of the `for` loop block.
- If `DangerousRelease` is used instead of `Dispose` in `AppleCertificatePal.DisposeTempKeychain` then the same test code above actually fails with the following exception:
```
System.ObjectDisposedException: Safe handle has been closed.
Object name: 'SafeHandle'.
   at System.Runtime.InteropServices.SafeHandle.InternalRelease(Boolean disposeOrFinalizeOperation)
   at System.Runtime.InteropServices.SafeHandle.DangerousRelease()
   at Internal.Cryptography.Pal.AppleCertificatePal.DisposeTempKeychain()
   at Internal.Cryptography.Pal.AppleCertificatePal.Dispose()
   at System.Security.Cryptography.X509Certificates.X509Certificate.Reset()
   at System.Security.Cryptography.X509Certificates.X509Certificate2.Reset()
   at System.Security.Cryptography.X509Certificates.X509Certificate.Dispose(Boolean disposing)
   at System.Security.Cryptography.X509Certificates.X509Certificate.Dispose()
```

> and then have the collection import call Dispose() on the temp handle here (after the loop ends).

That's already happening in `ApplePkcs12CertLoader.Dispose`.

Reply:
Looks like the mismatched references could be caused by a missing `SafeTemporaryKeychainHandle.TrackItem` somewhere which causes additional release from `SafeTemporaryKeychainHandle.UntrackItem` (called from `SafeKeychainItemHandle.ReleaseHandle` which is itself called from `AppleCertificatePal.Dispose`).

Reply:
I think I untangled the mess. `X509MoveToKeychain` takes a certificate as an input parameter but it changes its keychain. While `SafeTemporaryKeychainHandle.TrackItem` was previously called on the input certificate it was no-op since the item had no associated keychain. However, by the time the handle is disposed, it has the associated keychain, and thus `SafeTemporaryKeychainHandle.UntrackItem` would release one too many references.

It is plausible that fixing this would make the `AppleCertificatePal._tempKeychain` logic entirely unnecessary.

Reply:
I updated the PR description to explain what is going on.